src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs

```diff
@@ -9,33 +9,48 @@ namespace OrchardCore.Admin
 {
     public class AdminAreaControllerRouteMapper : IAreaControllerRouteMapper
     {
-        private readonly string _defaultAreaPattern;
+        private const string _defaultAreaPattern = "{area}/{controller}/{action}/{id?}";
+        private readonly string _adminUrlPrefix;
 
         public int Order => -1000;
 
         public AdminAreaControllerRouteMapper(IOptions<AdminOptions> adminOptions)
         {
-            _defaultAreaPattern = adminOptions.Value.AdminUrlPrefix + "/{area}/{controller}/{action}/{id?}";
+            _adminUrlPrefix = adminOptions.Value.AdminUrlPrefix;
         }
 
         public bool TryMapAreaControllerRoute(IEndpointRouteBuilder routes, ControllerActionDescriptor descriptor)
         {
-            if (descriptor.ControllerName == "Admin" ||
-                descriptor.ControllerTypeInfo.GetCustomAttribute<AdminAttribute>() != null ||
-                descriptor.MethodInfo.GetCustomAttribute<AdminAttribute>() != null
-                )
+            var controllerAttribute = descriptor.ControllerTypeInfo.GetCustomAttribute<AdminAttribute>();
+            var actionAttribute = descriptor.MethodInfo.GetCustomAttribute<AdminAttribute>();
+
+            if (descriptor.ControllerName != "Admin" && controllerAttribute == null && actionAttribute == null)
+            {
+                return false;
+            }
+
+            string name = null;
+            var pattern = _defaultAreaPattern;
+
+            if (!string.IsNullOrWhiteSpace(actionAttribute?.Template))
+            {
+                name = actionAttribute.RouteName;
+                pattern = actionAttribute.Template;
+            }
+            else if (!string.IsNullOrWhiteSpace(controllerAttribute?.Template))
             {
-                routes.MapAreaControllerRoute(
-                    name: descriptor.DisplayName,
-                    areaName: descriptor.RouteValues["area"],
-                    pattern: _defaultAreaPattern.Replace("{action}", descriptor.ActionName),
-                    defaults: new { controller = descriptor.ControllerName, action = descriptor.ActionName }
-                );
-
-                return true;
+                name = controllerAttribute.RouteName;
+                pattern = controllerAttribute.Template;
             }
 
-            return false;
+            routes.MapAreaControllerRoute(
+                name: name ?? $"Admin Route for {descriptor.DisplayName}",
```

Comment:
all you should need is this

```
name ?? descriptor.DisplayName
```

Reply:
As I wrote above, I thought it's good to make it distinct from maps made by `DefaultAreaControllerRouteMapper`. But I'm not married to the idea so I can revert it.

Reply:
I think to avoid a possible breaking change, it is best to keep naming as is. The route in the attribute will clean things up, but I don't think we should avoid breaking change if we can help it.